Add Seesaw Binding

Frankenslag opened the pull request:
This is a PR to address the issue/request #559 from @andersonimes to provide a binding for the Adafruit Soil Sensor which uses the Adafruit Seesaw framework.

Adafruit Seesaw is a near-universal converter framework which allows you to add and extend hardware support to any I2C-capable microcontroller or microcomputer. Instead of getting separate I2C GPIO expanders, ADCs, PWM drivers, etc, seesaw can be configured to give a wide range of capabilities.

This binding provides an API which is close to the one provided by Adafruit themselves but also implements IGpioController so that available Gpio pins can be used in place of any 'on board' ones but using the standard IoT API.

This binding was developed using the Adafruit Seesaw breakout board which uses the ATSAMD09 and the default firmware exposes the following capabilities
- 3 x 12-bit ADC inputs
- 3 x 8-bit PWM outputs
- 7 x GPIO with selectable pullup or pulldown
- 1 x NeoPixel output (up to 340 pixels)
- 1 x EEPROM with 64 byte of NVM memory (handy for storing small access tokens or MAC addresses)
- 1 x Interrupt output that can be triggered by any of the accessories

krwq commented on src/devices/Seesaw/README.md:
```diff
@@ -0,0 +1,46 @@
+## Summary
+Adafruit Seesaw is a near-universal converter framework which allows you to add add and extend hardware support to any I2C-capable microcontroller or microcomputer. Instead of getting separate I2C GPIO expanders, ADCs, PWM drivers, etc, seesaw can be configured to give a wide range of capabilities.
+
+This binding procvides an Api which is close to the one provided by Adafruit themselves but also implements IGpioController so that available Gpio pins can be used in place of any 'on board' ones but using the standard IoT API.
```
```suggestion
This binding provides an Api which is close to the one provided by Adafruit themselves but also implements IGpioController so that available Gpio pins can be used in place of any 'on board' ones but using the standard IoT API.
```

Frankenslag replied:
Done in the latest PR update.

krwq commented on src/devices/Seesaw/README.md:
```diff
@@ -0,0 +1,46 @@
+
+## Note 
+
+When using Seesaw devices with a Raspberry Pi it has been observed that errors sometimes happen on the I2C bus. The nature of this error may be the 'clock stretching' bug or may just be that the breakout board cannot accomodate the default I2C speed.
```
Please add a link to `may be the 'clock stretching' bug`

Frankenslag replied:
Done in the latest PR update.

krwq commented on src/devices/Seesaw/SeesawAdc.cs:
```diff
@@ -0,0 +1,39 @@
+        /// </summary>
+        /// <param name="pin">The pin number in the devices numbering scheme.</param>
+        /// <returns>A value between 0..1023 that represents the analog value.</returns> 
+        public UInt16 AnalogRead(byte pin)
```
nit: `UInt16` => `ushort`

Frankenslag replied:
Done in the latest PR update.

krwq commented on src/devices/Seesaw/SeesawAdc.cs:
```diff
@@ -0,0 +1,39 @@
+        /// <returns>A value between 0..1023 that represents the analog value.</returns> 
+        public UInt16 AnalogRead(byte pin)
+        {
+            const int AdcConversionDelay = 500; //500 uSeconds delay between read and write to allow Adc conversion time.
```
Add `microseconds` somewhere in the name, then not sure if comment will be any more useful

Frankenslag replied:
Done in the latest PR update.

krwq commented on src/devices/Seesaw/SeesawAdc.cs:
```diff
@@ -0,0 +1,39 @@
+
+            buffer = Read(Seesaw_Module.Adc, Seesaw_Function.Adc_Channel_Offset + pin - 2, 2, AdcConversionDelay);
+
+            return((UInt16)(((UInt16) buffer[0] << 8) | buffer[1]));
```
Use one of these for conversion https://docs.microsoft.com/en-us/dotnet/api/system.buffers.binary.binaryprimitives?view=netstandard-2.1

krwq commented on src/devices/Seesaw/SeesawAdc.cs:
```diff
@@ -0,0 +1,39 @@
+
+            return((UInt16)(((UInt16) buffer[0] << 8) | buffer[1]));
+        }
+
```
nit: extra enter

Frankenslag replied:
Done in the latest PR update.

krwq replied:
seems not

Frankenslag replied:
I really believe that this is done now. Not sure how it got rolled back.

krwq commented on src/devices/Seesaw/SeesawBase.cs:
```diff
@@ -0,0 +1,176 @@
+    {
+        private const byte Sessaw_Hardware_Id = 0x55;
+
+        public I2cDevice I2cDevice {get; private set;}
```
```suggestion
        public I2cDevice I2cDevice { get; private set; }
```

Frankenslag replied:
Done in the latest PR update.

krwq commented on src/devices/Seesaw/SeesawBase.cs:
```diff
@@ -0,0 +1,176 @@
+        private const byte Sessaw_Hardware_Id = 0x55;
+
+        public I2cDevice I2cDevice {get; private set;}
+        private UInt32 _options;
```
```suggestion
        private uint _options;
```

Frankenslag replied:
Done in the latest PR update.

krwq commented on src/devices/Seesaw/SeesawBase.cs:
```diff
@@ -0,0 +1,176 @@
+        {
+            byte[] buffer = Read(Seesaw_Module.Status, Seesaw_Function.Status_Temp, 4, 1000);
+
+            return ((1.0F / (1UL << 16)) * (((UInt32)buffer[0] << 24) | ((UInt32)buffer[1] << 16) | ((UInt32)buffer[2] << 8) | (UInt32)buffer[3]));
```
If powwible use https://docs.microsoft.com/en-us/dotnet/api/system.buffers.binary.binaryprimitives?view=netstandard-2.1

Frankenslag replied:
Done in the latest PR update.

krwq commented on src/devices/Seesaw/SeesawBase.cs:
```diff
@@ -0,0 +1,176 @@
+        /// <summary>
+        /// Performs a soft reset of the SeeSaw module. 
+        /// </summary>
+        public void SWReset()
```
Expand names on public APIs

Frankenslag replied:
Done in the latest PR update.

krwq commented on src/devices/Seesaw/SeesawBase.cs:
```diff
@@ -0,0 +1,176 @@
+        {
+            byte[] buffer = Read(Seesaw_Module.Status, Seesaw_Function.Status_Version, 4);
+
+            return (((UInt32)buffer[0] << 24) | ((UInt32)buffer[1] << 16) | ((UInt32)buffer[2] << 8) | (UInt32)buffer[3]);
```
https://docs.microsoft.com/en-us/dotnet/api/system.buffers.binary.binaryprimitives?view=netstandard-2.1

Frankenslag replied:
Done in the latest PR update.

krwq commented on src/devices/Seesaw/SeesawBase.cs:
```diff
@@ -0,0 +1,176 @@
+        {
+            byte[] buffer = Read(Seesaw_Module.Status, Seesaw_Function.Status_Options, 4);
+
+            return (((UInt32)buffer[0] << 24) | ((UInt32)buffer[1] << 16) | ((UInt32)buffer[2] << 8) | (UInt32)buffer[3]);
```
https://docs.microsoft.com/en-us/dotnet/api/system.buffers.binary.binaryprimitives?view=netstandard-2.1

krwq replied:
also might be worth to create a function line ReadUint since you are repeating this pattern several times

Frankenslag replied:
Done in the latest PR update... I have also added some common functions.

krwq commented on src/devices/Seesaw/SeesawBase.cs:
```diff
@@ -0,0 +1,176 @@
+
+            Span<byte> buffer = data.Length < StackThreshold ? stackalloc byte[data.Length + 2] : new byte[data.Length + 2];
+
+            for (int i = 0; i < data.Length; buffer[i + 2] = data[i], i++) ;
```
put `buffer[i + 2] = data[i]` in the body of the `for` loop

Frankenslag replied:
Done in the latest PR update by changes to just use Span<byte> as input.

krwq commented on src/devices/Seesaw/SeesawBase.cs:
```diff
@@ -0,0 +1,176 @@
+
+            I2cDevice.Write(new byte[2] { (byte)moduleAddress, (byte)functionAddress });
+
+            Task.Delay(delay/1000).Wait();
```
```suggestion
            Task.Delay(delay / 1000).Wait();
```

Frankenslag replied:
Done in the latest PR update.

krwq commented on src/devices/Seesaw/SeesawBase.cs:
```diff
@@ -0,0 +1,176 @@
+        protected byte[] Read(Seesaw_Module moduleAddress, Seesaw_Function functionAddress, int length, short delay = 0)
+        {
+
+            byte[] retval = new byte[length];
```
this can stackalloc

krwq commented on src/devices/Seesaw/SeesawBase.cs:
```diff
@@ -0,0 +1,176 @@
+
+            byte[] retval = new byte[length];
+
+            I2cDevice.Write(new byte[2] { (byte)moduleAddress, (byte)functionAddress });
```
(untested)

```suggestion
            Span<byte> bytesToWrite = stackalloc byte[2] { (byte)moduleAddress, (byte)functionAddress };
            I2cDevice.Write(bytesToWrite);
```

Frankenslag replied:
@krwq 
Done in the latest PR update but I am not sure why you state untested.

krwq commented on src/devices/Seesaw/SeesawBase.cs:
```diff
@@ -0,0 +1,176 @@
+
+            I2cDevice.Read(retval);
+
+            return (retval);
```
```suggestion
            return retval;
```

Frankenslag replied:
Done in the latest PR update.

krwq commented on src/devices/Seesaw/SeesawBase.cs:
```diff
@@ -0,0 +1,176 @@
+        protected byte[] Read(Seesaw_Module moduleAddress, Seesaw_Function functionAddress, int length, short delay = 0)
+        {
+
+            byte[] retval = new byte[length];
```
better solution would be taking `Span<byte>` as an input to the function (return `void` and `length` is implied by `Span`)

Frankenslag replied:
@krwq 
 But that would add complexity to the callers of the function. So for example, the following

`return Read(moduleAddress, functionAddress, 1, delay)[0];`

would need to be replaced by...

```
 Span<byte> buffer = stackalloc byte[2];
 Read(moduleAddress, functionAddress, buffer, delay);
 return buffer[0];

```

krwq commented on src/devices/Seesaw/SeesawEeprom.cs:
```diff
@@ -0,0 +1,80 @@
+        /// <returns>The data byte representing the I2C address.</returns>
+        public byte GetI2CAddress()
+        {
+            return(ReadEEPromByte((byte)Seesaw_Function.Eeprom_I2cAddr));
```
```suggestion
            return ReadEEPromByte((byte)Seesaw_Function.Eeprom_I2cAddr);
```

Frankenslag replied:
Done in the latest PR update.

krwq replied:
not seeing this being included

Frankenslag replied:
I have made this private as it is really only used by another function to determine if the I2C  address has been changed.

krwq commented on src/devices/Seesaw/SeesawGpio.cs:
```diff
@@ -0,0 +1,150 @@
+{
+    public partial class Seesaw : IDisposable
+    {
+
```
nit: remove extra enter

Frankenslag replied:
Done in the latest PR update.

krwq commented on src/devices/Seesaw/SeesawGpio.cs:
```diff
@@ -0,0 +1,150 @@
+                throw new ArgumentException("Gpio pin must be within 0-63 range.");
+            }
+
+            SetGpioPinModeBulk((UInt64)(1 << pin), mode);
```
```suggestion
            SetGpioPinModeBulk((ulong)(1 << pin), mode);
```

Frankenslag replied:
Done in the latest PR update.

krwq commented on src/devices/Seesaw/SeesawGpio.cs:
```diff
@@ -0,0 +1,150 @@
+                throw new NotSupportedException($"The hardware on I2C Bus {I2cDevice.ConnectionSettings.BusId}, Address 0x{I2cDevice.ConnectionSettings.DeviceAddress:X2} does not support Adafruit SeeSaw GPIO functionality");
+            }
+
+            byte[] buffer = new byte[] { (byte)(pins >> 24), (byte)(pins >> 16), (byte)(pins >> 8), (byte)pins, (byte)(pins >> 56), (byte)(pins >> 48), (byte)(pins >> 40), (byte)(pins >> 32)};
```
https://docs.microsoft.com/en-us/dotnet/api/system.buffers.binary.binaryprimitives?view=netstandard-2.1

Frankenslag replied:
@krwq Unfortunately the byte array is not formatted as a ulong converted ty bytes but as two uints representing the low and high 32 bits and so there is no obvious bit conversion that I can see.
I have created functions for this conversion and documented the byte array layout and it will be available in the next update

Frankenslag replied:
Done in the latest PR update.

krwq commented on src/devices/Seesaw/SeesawGpio.cs:
```diff
@@ -0,0 +1,150 @@
+                throw new NotSupportedException($"The hardware on I2C Bus {I2cDevice.ConnectionSettings.BusId}, Address 0x{I2cDevice.ConnectionSettings.DeviceAddress:X2} does not support Adafruit SeeSaw GPIO functionality");
+            }
+
+            byte[] buffer = new byte[] { (byte)(pins >> 24), (byte)(pins >> 16), (byte)(pins >> 8), (byte)pins, (byte)(pins >> 56), (byte)(pins >> 48), (byte)(pins >> 40), (byte)(pins >> 32) };
```
https://docs.microsoft.com/en-us/dotnet/api/system.buffers.binary.binaryprimitives?view=netstandard-2.1

Frankenslag replied:
@krwq Unfortunately the byte array is not formatted as a ulong converted ty bytes but as two uints representing the low and high 32 bits and so there is no obvious bit conversion that I can see.
I have created functions for this conversion and documented the byte array layout and it will be available in the next update

krwq commented on src/devices/Seesaw/SeesawGpio.cs:
```diff
@@ -0,0 +1,150 @@
+
+            byte[] buffer = Read(Seesaw_Module.Gpio, Seesaw_Function.Gpio_Bulk, 8);
+
+            return ((((UInt64)buffer[4] << 56) | ((UInt64)buffer[5] << 48) | ((UInt64)buffer[6] << 40) | ((UInt64)buffer[7] << 32) | ((UInt64)buffer[0] << 24) | ((UInt64)buffer[1] << 16) | ((UInt64)buffer[2] << 8) | buffer[3]) & pins);
```
https://docs.microsoft.com/en-us/dotnet/api/system.buffers.binary.binaryprimitives?view=netstandard-2.1

Frankenslag replied:
@krwq Unfortunately the byte array is not formatted as a ulong converted ty bytes but as two uints representing the low and high 32 bits and so there is no obvious bit conversion that I can see.
I have created functions for this conversion and documented the byte array layout and it will be available in the next update

krwq commented on src/devices/Seesaw/SeesawGpio.cs:
```diff
@@ -0,0 +1,150 @@
+        /// </summary>
+        /// <param name="pins">A 64bit integer containing 1 bit for each pin. If a bit is set to 1 then the interrupt is enabled for the associated pin..</param>
+        /// <param name="enable">A boolean value that indicates that interrupts are enabled when true or disabled when false.</param>
+        public void SetGpioInterrupts(UInt32 pins, bool enable)
```
is info about interrupt available through external pin or is there some function missing here how to use it?

krwq replied:
Possibly you should give a way to read Gpio_Intflag

Frankenslag replied:
@krwq 

Pin 8, on the breakout board, also labelled IRQ, is signalled when  an enabled GPIO pin changes state. I have added a function that reads the interruppt status and then clears the interrupt and this will be available in the next PR

krwq commented:
@Frankenslag is this ready for re-review or are you still making some changes here?

Frankenslag commented:
> @Frankenslag is this ready for re-review or are you still making some changes here?

@krwq 

I think tha I Have addressed the comments so please review. The latest PR.

krwq commented on src/devices/Seesaw/README.md:
```diff
@@ -0,0 +1,46 @@
+﻿# Adafruit Seesaw
+
+## Summary
+Adafruit Seesaw is a near-universal converter framework which allows you to add add and extend hardware support to any I2C-capable microcontroller or microcomputer. Instead of getting separate I2C GPIO expanders, ADCs, PWM drivers, etc, seesaw can be configured to give a wide range of capabilities.
```
nit: enter after `## Summary`

krwq commented on src/devices/Seesaw/README.md:
```diff
@@ -0,0 +1,46 @@
+
+## References 
+
+Please see the datasheet :-
```
`:-`?

krwq commented on src/devices/Seesaw/SeesawAdc.cs:
```diff
@@ -0,0 +1,35 @@
+
+            if (!HasModule(Seesaw_Module.Gpio))
+            {
+                throw new NotSupportedException($"The hardware on I2C Bus {I2cDevice.ConnectionSettings.BusId}, Address 0x{I2cDevice.ConnectionSettings.DeviceAddress:X2} does not support Adafruit SeeSaw ADC functionality");
```
I think this should be InvalidOperationException

Frankenslag replied:
Have looked for all places that this is and changed the exception to InvalidOperationException in the last PR update

krwq commented on src/devices/Seesaw/SeesawAdc.cs:
```diff
@@ -0,0 +1,35 @@
+
+            if (pin < 2 || pin > 5)
+            {
+                throw new ArgumentException("ADC pin must be within 2-5 range.");
```
ArgumentOutOfRangeException

Frankenslag replied:
Have looked for all places that this is and changed the exception to ArgumentOutOfRangeException in the last PR update

krwq commented on src/devices/Seesaw/SeesawAdc.cs:
```diff
@@ -0,0 +1,35 @@
+                throw new ArgumentException("ADC pin must be within 2-5 range.");
+            }
+
+            return(BinaryPrimitives.ReadUInt16LittleEndian(Read(Seesaw_Module.Adc, Seesaw_Function.Adc_Channel_Offset + pin - 2, 2, AdcConversionDelayMicroseconds)));
```
nit: space after return, outer brackets not needed

krwq commented on src/devices/Seesaw/SeesawBase.cs:
```diff
@@ -0,0 +1,172 @@
+{
+    public partial class Seesaw : IDisposable
+    {
+        private const byte Sessaw_Hardware_Id = 0x55;
```
```suggestion
        private const byte Seesaw_Hardware_Id = 0x55;
```

krwq replied:
btw if this is I2c address then it should be called `DefaultI2cAddress` to follow other devices

Frankenslag replied:
It's not an I2C address it a hardware identifier so that you can be sure you are talking to a Seesaw device. Underscores removed

krwq commented on src/devices/Seesaw/SeesawBase.cs:
```diff
@@ -0,0 +1,172 @@
+        /// <param name="moduleAddress">An Seesaw_Module enum that represents the module that we are writing to.</param>
+        /// <param name="functionAddress">An Seesaw_Function enum that represents the Seesaw function to be called.</param>
+        /// <param name="data">The Span containg data to be send as a parameter or parameters to the Seesaw device.</param>
+        protected void Write(Seesaw_Module moduleAddress, Seesaw_Function functionAddress, Span<byte> data)
```
```suggestion
        protected void Write(Seesaw_Module moduleAddress, Seesaw_Function functionAddress, ReadOnlySpan<byte> data)
```

krwq commented on src/devices/Seesaw/SeesawBase.cs:
```diff
@@ -0,0 +1,172 @@
+
+            Span<byte> buffer = data.Length < StackThreshold ? stackalloc byte[data.Length + 2] : new byte[data.Length + 2];
+
+            buffer[0] = (byte) moduleAddress;
```
nit: no space between cast operator and operand

krwq replied:
(same below)

Frankenslag replied:
Yes - for some reason Visual studio wasn't applying the .editorconfig suggestions. Hopefully I have found all occurrences and corrected in the latest PR update

krwq commented on src/devices/Seesaw/SeesawBase.cs:
```diff
@@ -0,0 +1,172 @@
+        /// <param name="length">The number of bytes that are expected to be returned from the Seesaw device.</param>
+        /// <param name="delay">The delay between sending the function all and readinng the data in microseconds.</param>
+        /// <returns>Returns the byte array representaing values from the Seesaw device.</returns>
+        protected byte[] Read(Seesaw_Module moduleAddress, Seesaw_Function functionAddress, int length, short delay = 0)
```
delay should contain units in name

krwq commented on src/devices/Seesaw/SeesawBase.cs:
```diff
@@ -0,0 +1,172 @@
+
+            I2cDevice.Write(bytesToWrite);
+
+            Task.Delay(delay / 1000).Wait();
```
use [DelayHelper](https://github.com/dotnet/iot/blob/master/src/devices/Common/System/Device/DelayHelper.cs)

krwq commented on src/devices/Seesaw/SeesawGpio.cs:
```diff
@@ -0,0 +1,201 @@
+                    Write(Seesaw_Module.Gpio, Seesaw_Function.Gpio_Bulk_Clr, pinArray);
+                    break;
+            }
+
```
nit: remove enter

krwq commented on src/devices/Seesaw/SeesawGpio.cs:
```diff
@@ -0,0 +1,201 @@
+        /// <returns>A ulong representing the 64 Gpio pins</returns>
+        private ulong PinArrayToPins(byte[] pinArray)
+        {
+            return (((ulong) pinArray[4] << 56) | ((ulong) pinArray[5] << 48) | ((ulong) pinArray[6] << 40) | ((ulong) pinArray[7] << 32) | ((ulong) pinArray[0] << 24) | ((ulong) pinArray[1] << 16) | ((ulong) pinArray[2] << 8) | (ulong) pinArray[3]);
```
remove spaces between cast operator and operand

krwq commented on src/devices/Seesaw/SeesawGpio.cs:
```diff
@@ -0,0 +1,201 @@
+        /// <returns>A byte array to write to a Seesaw device</returns>
+        private byte[] PinsToPinArray(ulong pins)
+        {
+            return new byte[] { (byte) (pins >> 24), (byte) (pins >> 16), (byte) (pins >> 8), (byte) pins, (byte) (pins >> 56), (byte) (pins >> 48), (byte) (pins >> 40), (byte) (pins >> 32) };
```
ditto

krwq commented on src/devices/Seesaw/SeesawGpioController.cs:
```diff
@@ -0,0 +1,233 @@
+{
+    public class SeesawGpioController : IGpioController
+    {
+
```
nit: extra enter

krwq commented on src/devices/Seesaw/SeesawGpioController.cs:
```diff
@@ -0,0 +1,233 @@
+
+            _openPins = new Dictionary<int, PinMode>();
+        }
+
```
nit: one enter between methods

krwq commented on src/devices/Seesaw/SeesawGpioController.cs:
```diff
@@ -0,0 +1,233 @@
+                throw new InvalidOperationException("Can not set a value to a pin that is not open.");
+            }
+
+            _seesawDevice.WriteGpioDigital((byte)pinNumber, (value==PinValue.High));
```
```suggestion
            _seesawDevice.WriteGpioDigital((byte)pinNumber, value == PinValue.High);
```

krwq commented on src/devices/Seesaw/SeesawGpioController.cs:
```diff
@@ -0,0 +1,233 @@
+
+        private void Dispose(bool disposing)
+        {
+            foreach(int pinNumber in _openPins.Keys)
```
```suggestion
            foreach (int pinNumber in _openPins.Keys)
```

krwq commented on src/devices/Seesaw/SeesawGpioController.cs:
```diff
@@ -0,0 +1,233 @@
+            {
+                ClosePin(pinNumber);
+            }
+            _openPins.Clear();
```
nit: enter

krwq commented on src/devices/Seesaw/SeesawGpioController.cs:
```diff
@@ -0,0 +1,233 @@
+        {
+            Dispose(true);
+        }
+
```
nit: enter

krwq commented on src/devices/Seesaw/SeesawModules.cs:
```diff
@@ -0,0 +1,66 @@
+{
+   public partial class Seesaw : IDisposable
+   {
+        public enum Seesaw_Module : Byte
```
remove underscore from all names here

krwq commented on src/devices/Seesaw/SeesawModules.cs:
```diff
@@ -0,0 +1,66 @@
+
+namespace Iot.Device.Seesaw
+{
+   public partial class Seesaw : IDisposable
```
rename the namespace or the class name, otherwise it will be really annoying to use (you will need to write both each time this is used)

Frankenslag replied:
@krwq 
Isn't this what the other devices have done though for example -

```
namespace Iot.Device.Mcp3008
{
    /// <summary>
    /// MCP3008 Analog to Digital Converter (ADC)
    /// </summary>
    public class Mcp3008 : IDisposable
    {
```

krwq commented on src/devices/Seesaw/SeesawModules.cs:
```diff
@@ -0,0 +1,66 @@
+            Gpio_Intflag = 0x0A,
+            Gpio_Pullenset = 0x0B,
+            Gpio_Pullenclr = 0x0C,
+
```
extra emter

krwq commented on src/devices/Seesaw/SeesawTouch.cs:
```diff
@@ -0,0 +1,27 @@
+    public partial class Seesaw : IDisposable
+    {
+        /// <summary>
+        /// Reads the analog value on an capacitive touch-enabled pin.. 
```
```suggestion
        /// Reads the analog value on an capacitive touch-enabled pin.
```

krwq commented on src/devices/Seesaw/SeesawTouch.cs:
```diff
@@ -0,0 +1,27 @@
+        /// <summary>
+        /// Reads the analog value on an capacitive touch-enabled pin.. 
+        /// </summary>
+        /// <param name="pinId">>The number of the pin to read.</param>
```
```suggestion
        /// <param name="pinId">The number of the pin to read.</param>
```

krwq commented on src/devices/Seesaw/SeesawTouch.cs:
```diff
@@ -0,0 +1,27 @@
+        /// <returns>An analogue value betweeen 0 and 1023 that represents the capacitance read from the pin.</returns>
+        public ushort TouchRead(byte pinId)
+        {
+            if(!HasModule(Seesaw_Module.Touch))
```
Ok, please fix the style in this PR, refer to https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md

I'll re-review once this is fixed

krwq commented on src/devices/Seesaw/SeesawBase.cs:
```diff
@@ -0,0 +1,161 @@
+        /// <returns>Returns the byte array representaing values from the Seesaw device.</returns>
+        protected byte[] Read(SeesawModule moduleAddress, SeesawFunction functionAddress, int length, short readDelayMicroSeconds = 0)
+        {
+
```
nit: enter

krwq commented on src/devices/Seesaw/SeesawEeprom.cs:
```diff
@@ -0,0 +1,77 @@
+        /// with the host device and dispose the current I2cDevice.
+        /// </summary>
+        /// <param name="i2cAddress">The new I2C address to be used.</param>
+        public void SetI2CAddress(byte i2cAddress)
```
`I2C` => `I2c`

krwq commented on src/devices/Seesaw/SeesawEeprom.cs:
```diff
@@ -0,0 +1,77 @@
+        /// Read the address configured to be used as the I2C address.
+        /// </summary>
+        /// <returns>The data byte representing the I2C address.</returns>
+        private byte GetI2CAddress() => ReadEEPromByte((byte)SeesawFunction.EepromI2cAddr);
```
ditto

krwq commented on src/devices/Seesaw/samples/Volume.cs:
```diff
@@ -0,0 +1,80 @@
+
+public class Volume
+{
+
```
nit: enter

krwq commented on src/devices/Seesaw/samples/Volume.cs:
```diff
@@ -0,0 +1,80 @@
+        return (int) value;
+    }
+
+    public void Dispose()
```
not needed since you're not implementing IDisposable

krwq commented on src/devices/Seesaw/samples/Volume.cs:
```diff
@@ -0,0 +1,80 @@
+    public static Volume EnableVolume(Seesaw seesawDevice)
+    {
+        var volume = new Volume(seesawDevice);
+        volume.Init();
```
why not just put init in the constructor?

krwq commented on src/devices/Seesaw/samples/Volume.cs:
```diff
@@ -0,0 +1,80 @@
+        _lastValue = GetVolumeValue();
+    }
+
+    public (bool update, int value) GetSleepforVolume(int sleep)
```
```suggestion
    public (bool update, int value) GetSleepForVolume(int sleep)
```

krwq commented:
LGTM with some comments, please don't force push new changes since it's hard to tell what changed

Frankenslag commented:
> LGTM with some comments, please don't force push new changes since it's hard to tell what changed

I have pushed the latest update addressing the minor issues you mentioned above. I believe that the push was forced as I had rebased onto the latest master. Not sure though. 

The Volume.cs code that you mention above has been amended but this is a copy taken of the 'more blinking lights' sample to show that the Seesaw device can do the same job as the on-board RaspberryPi io.

I have also changed the IGpioController to GpioDriver as mentioned in PR #684 and modified the sample to show it working. 

Hope all is good and sorry for the quick turnaround.

krwq commented:
Thanks for fixing it up @Frankenslag
